File: platform/src/main/java/org/stellar/anchor/platform/apiclient/CustodyApiClient.java
package org.stellar.anchor.platform.apiclient;

import static org.stellar.anchor.util.OkHttpUtil.TYPE_JSON;

import com.google.gson.Gson;
import java.io.IOException;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import org.apache.commons.lang3.StringUtils;
import org.springframework.http.HttpStatus;
import org.stellar.anchor.api.custody.CreateCustodyTransactionRequest;
import org.stellar.anchor.api.custody.CreateTransactionPaymentResponse;
import org.stellar.anchor.api.custody.CustodyExceptionResponse;
import org.stellar.anchor.api.custody.GenerateDepositAddressResponse;
import org.stellar.anchor.api.exception.CustodyException;
import org.stellar.anchor.api.exception.InvalidConfigException;
import org.stellar.anchor.auth.AuthHelper;
import org.stellar.anchor.platform.config.CustodyApiConfig;
import org.stellar.anchor.util.AuthHeader;
import org.stellar.anchor.util.GsonUtils;

public class CustodyApiClient {

  private static final Gson gson = GsonUtils.getInstance();

  private static final String CREATE_TRANSACTION_URL_FORMAT = "/transactions";
  private static final String GENERATE_DEPOSIT_ADDRESS_URL_FORMAT =
      "/transactions/payments/assets/%s/address";
  private static final String CREATE_TRANSACTION_PAYMENT_URL_FORMAT = "/transactions/%s/payments";

  private final OkHttpClient httpClient;
  private final AuthHelper authHelper;
  private final CustodyApiConfig custodyApiConfig;

  public CustodyApiClient(
      OkHttpClient httpClient, AuthHelper authHelper, CustodyApiConfig custodyApiConfig) {
    this.httpClient = httpClient;
    this.authHelper = authHelper;
    this.custodyApiConfig = custodyApiConfig;
  }

  public void createTransaction(CreateCustodyTransactionRequest transactionRequest)
      throws CustodyException, InvalidConfigException {
    Request request =
        getRequestBuilder()
            .url(custodyApiConfig.getBaseUrl() + CREATE_TRANSACTION_URL_FORMAT)
            .post(RequestBody.create(gson.toJson(transactionRequest).getBytes(), TYPE_JSON))
            .build();
    doRequest(request);
  }

  public GenerateDepositAddressResponse generateDepositAddress(String assetId)
      throws CustodyException, InvalidConfigException {
    Request request =
        getRequestBuilder()
            .url(
                custodyApiConfig.getBaseUrl()
                    + String.format(GENERATE_DEPOSIT_ADDRESS_URL_FORMAT, assetId))
            .post(RequestBody.create(StringUtils.EMPTY, null))
            .build();
    String responseBody = doRequest(request);
    return gson.fromJson(responseBody, GenerateDepositAddressResponse.class);
  }

  public CreateTransactionPaymentResponse createTransactionPayment(String txnId, String requestBody)
      throws CustodyException, InvalidConfigException {
    final String url =
        custodyApiConfig.getBaseUrl() + String.format(CREATE_TRANSACTION_PAYMENT_URL_FORMAT, txnId);

    Request request =
        getRequestBuilder()
            .url(url)
            .post(RequestBody.create(gson.toJson(requestBody).getBytes(), TYPE_JSON))
            .build();

    return gson.fromJson(doRequest(request), CreateTransactionPaymentResponse.class);
  }

  private Request.Builder getRequestBuilder() throws InvalidConfigException {
    Request.Builder requestBuilder = new Request.Builder();
    AuthHeader<String, String> authHeader = authHelper.createCustodyAuthHeader();
    return authHeader == null
        ? requestBuilder
        : requestBuilder.header(authHeader.getName(), authHeader.getValue());
  }

  private String doRequest(Request request) throws CustodyException {
    try (Response response = httpClient.newCall(request).execute()) {
      ResponseBody responseBody = response.body();
      String responseBodyJson = null;

      if (responseBody != null) {
        responseBodyJson = responseBody.string();
      }

      if (HttpStatus.valueOf(response.code()).is2xxSuccessful()) {
        return responseBodyJson;
      } else {
        String rawMessage =
            gson.fromJson(responseBodyJson, CustodyExceptionResponse.class).getRawErrorMessage();
        if (rawMessage != null) {
          throw new CustodyException(rawMessage, response.code());
        } else {
          throw new CustodyException(responseBodyJson, response.code());
        }
      }
    } catch (IOException e) {
      throw new CustodyException(e);
    }
  }
}

File: lib-util/src/main/kotlin/org/stellar/anchor/client/CustodyApiClient.kt
package org.stellar.anchor.client

import com.google.gson.reflect.TypeToken
import java.security.PrivateKey
import org.stellar.anchor.api.custody.CreateCustodyTransactionRequest
import org.stellar.anchor.api.custody.GenerateDepositAddressResponse
import org.stellar.anchor.util.RSAUtil

class CustodyApiClient(
  private val endpoint: String,
  private val jwt: String,
  private val webhookPrivateKey: PrivateKey
) : SepClient() {

  fun generateDepositAddress(asset: String): GenerateDepositAddressResponse {
    val url = "$endpoint/assets/$asset/addresses"
    val responseBody = httpPost(url, mapOf<String, String>(), jwt)
    return gson.fromJson(responseBody, GenerateDepositAddressResponse::class.java)
  }

  fun createTransaction(custodyTransaction: CreateCustodyTransactionRequest) {
    val url = "$endpoint/transactions"
    val type = object : TypeToken<Map<String?, *>?>() {}.type
    val requestBody: Map<String, Any> = gson.fromJson(gson.toJson(custodyTransaction), type)
    httpPost(url, requestBody, jwt)
  }

  fun createTransactionPayment(transactionId: String) {
    val url = "$endpoint/transactions/$transactionId/payments"
    httpPost(url, mapOf<String, String>(), jwt)
  }

  fun sendWebhook(webhookRequest: String) {
    val webhookSignature = RSAUtil.sign(webhookRequest, webhookPrivateKey)
    httpPost("$endpoint/webhook", webhookRequest, mapOf("fireblocks-signature" to webhookSignature))
  }
}
